src/input/HidApiInputDriver.cc

```diff
@@ -178,36 +233,99 @@ void HidApiInputDriver::hidapi_input(hid_device* hid_dev)
     unsigned char buf[BUFLEN];
     unsigned int len;
     while ((len = hid_read(hid_dev, buf, BUFLEN)) > 0) {
+		hidapi_log_input(buf, len);
         (this->*(dev->axis_decoder))(buf, len);
         (this->*(dev->button_decoder))(buf, len);
     }
     hid_close(hid_dev);
 }
 
+std::pair<hid_device *, const struct device_id *> HidApiInputDriver::enumerate() const
+{
+	hid_device *ret_hid_dev = nullptr;
+	const struct device_id *ret_dev = nullptr;
+
+	HIDAPI_LOG("Enumerating HID devices...");
+	struct hid_device_info *info = hid_enumerate(0, 0);
+	for (;info != nullptr;info = info->next) {
+		HIDAPI_LOGP("D: %04x:%04x | path = %s, serial = %s, manufacturer = %s, product = %s",
+				info->vendor_id % info->product_id % info->path
+				% to_string(info->serial_number)
+				% to_string(info->manufacturer_string)
+				% to_string(info->product_string));
+		const device_id *dev = match_device(info);
+		if (!dev) {
+			continue;
+		}
+
+		hid_device *hid_dev;
+
+		HIDAPI_LOGP("P: %04x:%04x | %s", info->vendor_id % info->product_id % info->path);
+		hid_dev = hid_open_path(info->path);
+
+		if (!hid_dev) {
+			HIDAPI_LOGP("O: %04x:%04x | %s", info->vendor_id % info->product_id % to_string(info->serial_number));
+			hid_dev = hid_open(info->vendor_id, info->product_id, info->serial_number);
+			if (!hid_dev) {
+				continue;
+			}
+		}
+
+		HIDAPI_LOGP("R: %04x:%04x | %s", info->vendor_id % info->product_id % to_string(info->serial_number));
+		unsigned char buf[BUFLEN];
+		const int len = hid_read_timeout(hid_dev, buf, BUFLEN, 100);
+		HIDAPI_LOGP("?: %d", len);
+
+		if (len < 0) {
+			HIDAPI_LOGP("E: %s", to_string(hid_error(hid_dev)));
+			hid_close(hid_dev);
+			continue;
+		}
+
+		ret_dev = dev;
+		ret_hid_dev = hid_dev;
+		break;
+	}
+	hid_free_enumeration(info);
+	HIDAPI_LOGP("Done enumerating (status = %s).", (ret_hid_dev != nullptr ? "ok" : "failed"));
+	return std::make_pair(ret_hid_dev, ret_dev);
```

Comment:
Technically simplifiable to `return {ret_hid_dev, ret_dev};` I think

Reply:
Oh, nice, I did not know C++11 already handles that too...

Reply:
At least I _think_ it does; I sometimes confuse C++11 and C++14..

Reply:
Works locally, I guess we'll see what ancient Travis has to say :-).